saleor/checkout/complete_checkout.py

```diff
@@ -0,0 +1,190 @@
+
+from django.core.exceptions import ValidationError
+from django.db import transaction
+
+from ..account.error_codes import AccountErrorCode
+from ..account.models import User
+from ..core.exceptions import InsufficientStock
+from ..core.taxes import TaxError
+from ..core.utils.url import validate_storefront_url
+from ..discount import DiscountInfo
+from ..discount.models import NotApplicable
+from ..graphql.checkout.utils import (  # TODO move from graphql
+    clean_checkout_payment,
+    clean_checkout_shipping,
+)
+from ..order import models as order_models
+from ..order.models import Order
+from ..payment import PaymentError, gateway
+from ..payment.models import Payment, Transaction
+from ..payment.utils import store_customer_id
+from . import models
+from .calculations import calculate_checkout_total_with_gift_cards
+from .error_codes import CheckoutErrorCode
+from .utils import abort_order_data, create_order, get_order, prepare_order_data
+
+
+@transaction.atomic
+def prepare_checkout(checkout: models.Checkout, discounts, tracking_code, redirect_url):
+    lines = list(checkout)
+
+    clean_checkout_shipping(checkout, lines, discounts, CheckoutErrorCode)
+    clean_checkout_payment(checkout, lines, discounts, CheckoutErrorCode)
+
+    payment = checkout.get_last_active_payment()
+
+    validate_payment_amount(discounts, payment, checkout)
+
+    if redirect_url:
+        try:
+            validate_storefront_url(redirect_url)
+        except ValidationError as error:
+            raise ValidationError(
+                {"redirect_url": error}, code=AccountErrorCode.INVALID.value
+            )
+
+    to_update = []
+    if redirect_url and redirect_url != checkout.redirect_url:
+        checkout.redirect_url = redirect_url
+        to_update.append("redirect_url")
+
+    if tracking_code and tracking_code != checkout.tracking_code:
+        checkout.tracking_code = tracking_code
+        to_update.append("tracking_code")
+
+    if to_update:
+        to_update.append("last_change")
+        checkout.save(update_fields=to_update)
+
+
+@transaction.atomic
+def convert_checkout_to_order(
+    checkout: models.Checkout, order_data: dict, user: Optional[User]
+) -> order_models.Order:
+    order = create_order(
+        checkout=checkout, order_data=order_data, user=user,  # type: ignore
+    )
+    # remove checkout after order is successfully created
+    checkout.delete()
+    return order
+
+
+def validate_payment_amount(discounts, payment, checkout):
+    checkout_total = calculate_checkout_total_with_gift_cards(checkout, discounts)
+    if payment.total != checkout_total.gross.amount:
+        gateway.payment_refund_or_void(payment)
+        raise ValidationError(
+            "Payment does not cover all checkout value.",
+            code=CheckoutErrorCode.CHECKOUT_NOT_FULLY_PAID.value,
+        )
+
+
+def get_order_data(checkout: models.Checkout, discounts: List[DiscountInfo]) -> dict:
+    try:
+        with transaction.atomic():
```

Comment:
is this atomic needed here?

Reply:
So what do you propose? It prepares the data for order like locking the vouchers etc

Reply:
Keep in mind that if we do locking, anything we `SELECT FOR UPDATE` will be unlocked as soon as the transaction terminates.

Reply:
I agree with @tomaszszymanski129, I would expect this function to only return some data. Also, `prepare_order_data` seems to be used only in "checkout complete", so why isn't it in this module?

Reply:
@maarcingebala  @tomaszszymanski129  So what name do you propose?

Reply:
Hard to think of anything appropriate, maybe something like assemble_order? @maarcingebala any other idea?